infer/environment.cc

```diff
@@ -1587,7 +1587,7 @@ Environment::processBinding(core::Context ctx, const cfg::CFG &inWhat, cfg::Bind
                         }
                     }
                 } else if (!bind.value.isSynthetic()) {
-                    if (castType.isUntyped()) {
+                    if (castType.isUntyped() && bind.bind.variable != cfg::LocalRef::selfVariable()) {
```

Comment:
Non-blocking: might be worth mentioning why we have this selfVariable check here (e.g.: `self` is not actually assignable in normal Ruby syntax, so if the target of this binding is `<self>` we know that this `T.cast` is actually a `T.bind` in disguise, but we've already desugared `T.bind` to `T.cast` so we can't simply check that on the `c.cast` directly)